### DISCUSSION TITLE
ESP32: Possible scripter memory leak

### DISCUSSION THREAD

**Discussion by UBWH** [arendst/Tasmota#20430] [Support request]
<img width="651" alt="image" src="https://github.com/arendst/Tasmota/assets/72185209/22a830f9-a9f8-4bfe-8115-63172cf34656">

There seems to be a memory leak in V3.1.0 - I suspect in the SCRIPT code??? (EDIT: or maybe the Ethernet code ???)

In the above plot (over 7 days) - the main drops occur when I display the >w0, >w2, ... Web pages defined in my SCRIPT

I am at a loss to workout what to do next.

The MQTT messages are still flowing out - even with only 4kB of heap left (thats how I know how much heap is left - and can plot on OpenHAB)

But  I can NOT login to the main web page. Browser just times out waiting for data.

EDIT: Browser timing out started when heap dropped below about 20 kB

Suggestions on how to fix this?

EDIT: Using WT32-ETH01, with 4 MB memory, and Ethernet port.

**Comment by Jason2866** [comment]
For finding the culprit it is necessary to rule out where to search. Use without Ethernet.
Since it happens when doing Display stuff, looks like a issue in scripter. Your code or the scripter implementation. Role back the scripter changes done to see if and which commit is causing.

**Comment by UBWH** [comment]
I have followed @Jason2866  's advice and stripped back almost 100% of my script and found a simple way to demonstrate the issue, which seems to have nothing to do with the dual network interfaces: just a SCRIPT problem.

**SETUP:**
* ESP32
* Tasmota V3.1.0, with USE_SCRIPT defined
* ETHERNET: OFF
* WIFI: ON

# Test Process
Simply fetched the MAIN Tasmota page, with WGET. Used this bash script:
```
#!/bin/bash
for (( c=1; c<=200; c++ ))
do
   echo "$c times"
  wget  http://10.1.1.7/ -o /dev/null
  sleep 5
done
```
# TEST 1 - SUCCESS
## Zero drop of heap over 400 web gets of main Tasmota page.
SCRIPT
```
>D   

>U

>E

>S

>R

>J  

>T

```
# TEST 2 - FAIL
## Heap dropped form 97->48kB over 400 web gets of main page. (ran bash script twice)
 = approx 0.12kB per web get
SCRIPT
```
>D   

>U

>E

>S

>R

>J  

>T

>w6 About

```

Here is the plot of the heap while the script  ran twice.
<img width="652" alt="HeapPlot" src="https://github.com/arendst/Tasmota/assets/72185209/f9530007-962e-4d89-9667-d94c1042e011">


I am on holidays and have no access to my Visual Studio to built a new Tasmota binary on the latest release to confirm this memory leak bug has not already been fixed.

Should I raise a new issue anyway?

**Reply by Jason2866** [reply]
No change the headline. Something like "Possible scripter memory leak" AND provide the script which causes the leak. There is a chance the script itself is to blame.

**Comment by UBWH** [comment]
Jason,

Headline updated.

**Script(s)**
I DID provide them. They are almost empty. See above

Adding `>w6 About` triggers the memory loss.

**Would like to add**
- during this testing never clicked the new `About` button on the main page. (But if I did, I assume it would display a blank page)

**Comment by Jason2866** [comment]
@gemu2015 Can you help?

**Reply by gemu2015** [reply]
ok, found the bug, will do pr in the next few days

**Reply by UBWH** [reply]
Have you made the PR yet? 
I would like to patch my build and give it a test.

**Reply by gemu2015** [reply]
yes, done

**Comment by UBWH** [comment]
Thanks
:-)

**Comment by UBWH** [comment]
Fixed: See This PR: https://github.com/arendst/Tasmota/pull/20473

**Comment by mi-hol** [comment]
> <img alt="image" width="651" src="https://private-user-images.githubusercontent.com/72185209/294994035-22a830f9-a9f8-4bfe-8115-63172cf34656.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcxOTk1NDQsIm5iZiI6MTczNzE5OTI0NCwicGF0aCI6Ii83MjE4NTIwOS8yOTQ5OTQwMzUtMjJhODMwZjktYTlmOC00YmZlLTgxMTUtNjMxNzJjZjM0NjU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE4VDExMjA0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3M2YzMzk4MmVkMjY3ZWM5NzEwN2UwYmM4OTI5ODJhZWUyMTU2YzNhMWYzY2QwMDBhZWU3M2RiYzE2MmVmNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YlXjysnmnv7TGNvPonmyrytp2O21urBUJDhF1R6xKxM">


@UBWH We are hunting for root cause of reboot after~5 Min in https://github.com/arendst/Tasmota/discussions/22833

Would you perhaps have a link to instructions  "how to monitor free heap" as shown in your graph ?

**Reply by sfromis** [reply]
The general approach is that Tasmota is publishing STATE payloads via MQTT every TelePeriod, which among other fields has a value like `"Heap":63`. Those payloads can be subscribed to, stored, and presented by some backend tool, names like InfluxDB and Grafana comes to mind.

Of course, this type of monitoring is for longer time trends, not very useful for short time crashes. Using `SetOption130` you can get Tasmota to include heap size in every console log message, which may work better as long as there is activity.

**Reply by mi-hol** [reply]
@sfromis actually using SetOption130  causes another issue on C6, see https://github.com/arendst/Tasmota/issues/22839
